openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,236 @@
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/kubelet/cm"
+)
+
+const (
+	PluginName                 = "autoscaling.openshift.io/ManagementCPUsOverride"
+	podWorkloadLabel           = "workload.openshift.io"
+	managedCapacityLabel       = "workload.openshift.io/cores"
+	containerCPUsAnnotation    = "io.openshift.workload.management.cpu"
+	containerCPUSetAnnotation  = "io.openshift.workload.management.cpuset"
+	namespaceAllowedAnnotation = "workload.openshift.io/allowed"
+	workloadTypeManagement     = "management"
+)
+
+var _ = initializer.WantsExternalKubeInformerFactory(&managementCPUsOverride{})
+var _ = admission.MutationInterface(&managementCPUsOverride{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managementCPUsOverride{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister corev1listers.NamespaceLister
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace: %v", PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isNamespaceAllowed(ns.Annotations) {
+		return nil
+	}
+
+	if !isPodManaged(pod.Labels) {
+		return nil
+	}
+
+	allContainers := append([]coreapi.Container{}, pod.Spec.InitContainers...)
+	allContainers = append(allContainers, pod.Spec.Containers...)
+	podQoSClass := getPodQoSClass(allContainers)
+
+	// we do not want to change guaranteed pods resource allocation, because it should be managed by
+	// relevant managers(CPU and memory) under the kubelet
+	if podQoSClass == coreapi.PodQOSGuaranteed {
+		klog.Warningf("Skip pod %q CPUs requests modifications because it has guaranteed QoS class", pod.Name)
+		return nil
+	}
+
+	podCopy := pod.DeepCopy()
+	updateContainerResources(podCopy, podQoSClass)
+
+	allContainers = append([]coreapi.Container{}, podCopy.Spec.InitContainers...)
```

Comment:
don't reuse variables for different pruposes. This is bad practice.

Reply:
done